src/openfermion/ops/_symbolic_operator.py

```diff
@@ -0,0 +1,520 @@
+#   Licensed under the Apache License, Version 2.0 (the "License");
+#   you may not use this file except in compliance with the License.
+#   You may obtain a copy of the License at
+#
+#       http://www.apache.org/licenses/LICENSE-2.0
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+
+"""SymbolicOperator is the base class for FermionOperator and QubitOperator"""
+import copy
+
+from openfermion.config import EQ_TOLERANCE
+
+
+class SymbolicOperatorError(Exception):
+    pass
+
+
+class SymbolicOperator(object):
+    """Base class for FermionOperator and QubitOperator.
+
+    A SymbolicOperator stores an object which represents a weighted
+    sum of terms; each term is a product of individual factors
+    of the form (`index`, `action`), where `index` is a nonnegative integer
+    and the possible values for `action` are determined by the subclass.
+    For instance, for the subclass FermionOperator, `action` can be 1 or 0,
+    indicating raising or lowering, and for QubitOperator, `action` is from
+    the set {'X', 'Y', 'Z'}.
+    The coefficients of the terms are stored in a dictionary whose
+    keys are the terms.
+    SymbolicOperators of the same type can be added or multiplied together.
+
+    Attributes:
+        actions (tuple): A tuple of objects representing the possible actions.
```

Comment:
I wonder if it would be better if everything other than the terms attribute were a private attribute that is accessed with getters/setters. I am not necessarily of that belief, but it is worth thinking about. @jarrodmcc do you have any thoughts?